incountry/validation/validate_model.py

```diff
@@ -1,83 +1,110 @@
+    return reduce(lambda agg, field_name: f"{agg} \n  {field_name} - extra fields not permitted", extra_fields, "")
+
+
+def get_validated_function_input_for_models(
+    function: Callable,
+    args: list,
+    kwargs: dict,
+    models: List[Union[BaseModel, Dict]],
+    forbid_extra_non_model_attributes: bool = True,
+):
+    class_instance = None
+    model_fields_set = set()
+    new_kwargs = dict(kwargs)
+    function_args_to_kwargs(function, args, new_kwargs)
+
+    if "self" in signature(function).parameters:
+        class_instance = args[0]
+        model_fields_set.add("self")
+
+    for model in models:
+        validation_exception = InputValidationException
+        res_model = model
+
+        if isinstance(model, dict):
+            if "condition" in model and not model["condition"](class_instance):
+                continue
+
+            res_model = model["model"]
+            validation_exception = model.get("validation_exception", InputValidationException)
 
-        function_args_to_kwargs(function, args, new_kwargs)
+        model_fields_set.update(res_model.__fields__.keys())
 
         validated_data_dict = get_validated_data(
             function=function,
             model=res_model,
-            exclude_to_dict=exclude_to_dict,
             validation_exception=validation_exception,
             **new_kwargs,
         )
 
-        exception_context_str = f"during {function.__qualname__}()"
-
         for key in new_kwargs:
             if key in validated_data_dict:
                 new_kwargs[key] = validated_data_dict[key]
-        try:
-            return function(**new_kwargs)
-        except StorageClientException as e:
-            if exception_context_str in str(e):
-                raise
-            raise StorageClientException(f"Validation failed {exception_context_str}") from e
-        except StorageServerException as e:
-            if exception_context_str in str(e):
-                raise
-            if isinstance(e, StorageCountryNotSupportedException):
-                raise StorageCountryNotSupportedException(
-                    message=f"server exception {exception_context_str}", country=e.country
-                )
-            raise e.__class__(
-                message=f"Server exception {exception_context_str}",
-                url=e.url,
-                status_code=e.status_code,
-                method=e.method,
-                scope=e.scope,
-            ) from e
-        except StorageException as e:
-            raise e
-        except Exception as e:
-            raise StorageException(f"{e.__class__.__qualname__} {exception_context_str}") from e
+
+    if forbid_extra_non_model_attributes and not set(new_kwargs.keys()).issubset(model_fields_set):
+        raise InputValidationException(
+            f"Validation failed during {function.__qualname__}():"
+            f"{format_extra_fields_error(set(new_kwargs.keys()).difference(model_fields_set))}"
```

Comment:
`dict_keys` implement `__sub__`, so we could simplify it without additional reallocation like this:
```suggestion
            f"{format_extra_fields_error(new_kwargs.keys() - model_fields_set)}"
```

Reply:
[issubset](https://docs.python.org/3/library/stdtypes.html#frozenset.issubset) is also implemented by `dict_keys`